File: java/src/string/aho/AhoTest.java
package string.aho;

import org.junit.Assert;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Random;

public class AhoTest {
    @Test
    public void testContainsSingle() {
        Assert.assertArrayEquals(new boolean[]{true}, Aho.contains("ab", new String[]{"a"}));
        Assert.assertArrayEquals(new boolean[]{true}, Aho.contains("ab", new String[]{"ab"}));
        Assert.assertArrayEquals(new boolean[]{true}, Aho.contains("ab", new String[]{"b"}));
        Assert.assertArrayEquals(new boolean[]{true}, Aho.contains("abc", new String[]{"bc"}));
        Assert.assertArrayEquals(new boolean[]{true}, Aho.contains("abc", new String[]{"c"}));
    }

    @Test
    public void testContainsExamples() {
        Assert.assertArrayEquals(
            new boolean[]{true, true, true},
            Aho.contains("xxxabc",
            new String[]{"abc", "bc", "c"}));

        Assert.assertArrayEquals(
            new boolean[]{true, true, true, true, true, false},
            Aho.contains("ababa",
                new String[]{"baba", "ab", "ababa", "a", "b", "babab"}));

        Assert.assertArrayEquals(
            new boolean[]{false, true},
            Aho.contains("babb",
                new String[]{"baba", "ab"}));

        Assert.assertArrayEquals(
            new boolean[]{true, true, true, true, false},
            Aho.contains("ahishers",
                new String[]{"he", "she", "hers", "his", "him"}));
    }

    @Test
    public void testRandomCases() {
        var random = new Random(0);
	    final var max = 20;

        for (var test = 0; test < 10_000; test++) {
            var text = generateString(random, max, 1+random.nextInt(26));
            var strs = new String[1+random.nextInt(max)];
            var m = 1 + random.nextInt(26);
            for (var i = 0; i < strs.length; i++) {
                strs[i] = generateString(random, max, m);
            }

            strs = removeDuplicates(strs);
            var got = Aho.contains(text, strs);
            var want = containsBruteForce(text, strs);
            if (!Arrays.equals(got, want)) {
                Assert.fail(String.format("Aho.contains(%q, %q):\nwant %s\n got %s",
                    text, Arrays.toString(strs), Arrays.toString(want), Arrays.toString(got)));
            }
        }
    }

    private static boolean[] containsBruteForce(String text, String[] strs) {
        var ans = new boolean[strs.length];

        for (var i = 0; i < strs.length; i++) {
            ans[i] = text.contains(strs[i]);
        }

        return ans;
    }

    private static String generateString(Random random, int n, int m) {
        var str = new char[1+random.nextInt(n)];
        for (var i = 0; i < str.length; i++) {
            str[i] = (char)('a' + random.nextInt(m));
        }
        return new String(str);
    }

    private static String[] removeDuplicates(String[] strs) {
        var known = new HashSet<String>();
        var ans = new ArrayList<String>();
        for (var str : strs) {
            if (!known.contains(str)) {
                ans.add(str);
                known.add(str);
            }
        }
        return ans.toArray(new String[]{});
    }
}

File: go/string/aho/aho_test.go
package aho

import (
	"github.com/stretchr/testify/assert"
	"math/rand/v2"
	"reflect"
	"strings"
	"testing"
)

func TestContainsSingle(t *testing.T) {
	assert.Equal(t, []bool{true}, Contains("ab", []string{"a"}))
	assert.Equal(t, []bool{true}, Contains("ab", []string{"ab"}))
	assert.Equal(t, []bool{true}, Contains("ab", []string{"b"}))
	assert.Equal(t, []bool{true}, Contains("abc", []string{"bc"}))
	assert.Equal(t, []bool{true}, Contains("abc", []string{"c"}))
}

func TestContainsExamples(t *testing.T) {
	tests := []struct {
		text string
		strs []string
		want []bool
	}{
		{
			text: "xxxabc",
			strs: []string{"abc", "bc", "c"},
			want: []bool{true, true, true},
		},
		{
			text: "ababa",
			strs: []string{"baba", "ab", "ababa", "a", "b", "babab"},
			want: []bool{true, true, true, true, true, false},
		},
		{
			text: "babb",
			strs: []string{"baba", "ab"},
			want: []bool{false, true},
		},
		{
			text: "ahishers",
			strs: []string{"he", "she", "hers", "his", "him"},
			want: []bool{true, true, true, true, false},
		},
		{
			text: "cbbacbbdb",
			strs: []string{"bbbaabbbaaaaaaabaab", "aabbab", "bbb", "abbaaa", "bbababaabaaabaaa", "aabb",
				"a", "bbaba", "ba", "aaabaabbbbbbbbabbb"},
			want: []bool{false, false, false, false, false, false, true, false, true, false},
		},
	}

	for _, test := range tests {
		got := Contains(test.text, test.strs)
		if !reflect.DeepEqual(got, test.want) {
			t.Fatalf("Contains(%v, %v):\nwant %v\n got %v", test.text, test.strs, test.want, got)
		}
	}
}

func TestFindEndingsRandom(t *testing.T) {
	random := rand.New(rand.NewPCG(0, 0))
	const max = 20

	for test := 0; test < 10_000; test++ {
		text := generateString(random, max, 1+random.IntN(26))
		strs := make([]string, 1+random.IntN(max))
		m := 1 + random.IntN(26)
		for i, _ := range strs {
			strs[i] = generateString(random, max, m)
		}

		strs = removeDuplicates(strs)
		got, want := Contains(text, strs), containsBruteForce(text, strs)
		if !reflect.DeepEqual(got, want) {
			t.Fatalf("Contains(%v, %v):\nwant %v\n got %v", text, strs, want, got)
		}
	}
}

func containsBruteForce(text string, strs []string) (ans []bool) {
	for _, str := range strs {
		ans = append(ans, strings.Contains(text, str))
	}
	return
}

func generateString(random *rand.Rand, n, m int) string {
	str := make([]byte, 1+random.IntN(n))
	for i, _ := range str {
		str[i] = byte('a' + random.IntN(m))
	}
	return string(str)
}

func removeDuplicates(strs []string) (out []string) {
	known := map[string]bool{}
	for _, str := range strs {
		if !known[str] {
			out, known[str] = append(out, str), true
		}
	}
	return
}
